azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+    callback_called.wait()
+
+
 TestPipelineRootStagePipelineThreading.test_runs_callback_in_callback_thread = (
     _test_pipeline_root_runs_callback_in_callback_thread
 )
 TestPipelineRootStagePipelineThreading.test_runs_operation_in_pipeline_thread = (
     _test_pipeline_root_runs_operation_in_pipeline_thread
 )
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_connected_in_callback_thread = (
+    _test_pipeline_root_runs_on_connected_in_callback_thread
+)
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_disconnected_in_callback_thread = (
+    _test_pipeline_root_runs_on_disconnected_in_callback_thread
+)
+TestPipelineRootStagePipelineThreading.test_pipeline_root_runs_on_event_received_in_callback_thread = (
+    _test_pipeline_root_runs_on_event_received_in_callback_thread
+)
+
+pipeline_stage_test.add_base_pipeline_stage_tests(
+    cls=pipeline_stages_base.EnsureConnectionStage,
+    module=this_module,
+    all_ops=all_common_ops,
+    handled_ops=[
+        pipeline_ops_mqtt.MQTTPublishOperation,
+        pipeline_ops_mqtt.MQTTSubscribeOperation,
+        pipeline_ops_mqtt.MQTTUnsubscribeOperation,
+    ],
+    all_events=all_common_events,
+    handled_events=[],
+)
+
+fake_topic = "__fake_topic__"
+fake_payload = "__fake_payload__"
+ops_that_cause_connection = [
+    {
+        "op_class": pipeline_ops_mqtt.MQTTPublishOperation,
+        "op_init_kwargs": {"topic": fake_topic, "payload": fake_payload},
+    },
+    {"op_class": pipeline_ops_mqtt.MQTTSubscribeOperation, "op_init_kwargs": {"topic": fake_topic}},
+    {
+        "op_class": pipeline_ops_mqtt.MQTTUnsubscribeOperation,
+        "op_init_kwargs": {"topic": fake_topic},
+    },
+]
+
+
+@pytest.mark.parametrize(
+    "params",
+    ops_that_cause_connection,
+    ids=[x["op_class"].__name__ for x in ops_that_cause_connection],
+)
+@pytest.mark.describe("EnsureConnectionStage - .run_op() -- called with connect operation")
+class TestEnsureConnectionStageRunOp(object):
+    @pytest.fixture
+    def op(self, mocker, params):
+        op = params["op_class"](**params["op_init_kwargs"])
+        op.callback = mocker.MagicMock()
+        return op
+
+    @pytest.fixture
+    def stage(self, mocker):
+        stage = make_mock_stage(
+            mocker=mocker, stage_to_make=pipeline_stages_base.EnsureConnectionStage
+        )
+        stage.next.run_op = mocker.MagicMock()
+        return stage
+
+    @pytest.mark.it("Passes the operation down the pipline when the transport is already connected")
+    def test_operation_alrady_connected(self, params, op, stage):
+        stage.pipeline_root.connected = True
+
+        stage.run_op(op)
+
+        assert stage.next.run_op.call_count == 1
+        assert stage.next.run_op.call_args[0][0] == op
+
+    @pytest.mark.it(
+        "Sends a ConnectOperation instead of the op down the pipeline if the transport is not connected"
+    )
+    def test_sends_connect(self, params, op, stage):
```

Comment:
>params [](start = 33, length = 6)

If you put the parametrization directly on the fixture rather than the whole class, you wouldn't need to include the `params` args in all of these tests

Reply:
yes I would.  The runner requires params on the test function even if it's not used.  Otherwise it assume you forgot to include it and fails with error.

---
In reply to: [312670685](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r312670685) [](ancestors = 312670685)